test/utils/ruleset-test-harness.ts

```diff
@@ -0,0 +1,151 @@
+/* eslint-disable class-methods-use-this */
+/* eslint-disable no-underscore-dangle */
+import 'mocha';
+import {
+  Browser, Page, chromium, firefox, webkit,
+} from 'playwright';
+
+import {
+  expectFS, ExpectObserver, setupGlobals, expectGlobal,
+} from './mocha';
+
+export interface RulesetTestHarness {
+  setUp: (rules: any, dataLayer: any) => Promise<void>;
+  tearDown: () => Promise<void>;
+  execute: (action: (args: any[]) => void, args?: any[]) => Promise<void>;
+  popEvent: (timeoutMs?: number) => Promise<any[]>;
+}
+
+interface RulesetTestEnvironment {
+  name: string;
+  createTestHarness: (rules: any, dataLayer: any) => Promise<RulesetTestHarness>;
+  tearDown: () => Promise<void>;
+}
+
+const isBrowserTest = process.env.DLO_RUN_BROWSER_TESTS;
+const dloScriptSrc = process.env.PLAYWRIGHT_DLO_SCRIPT_SRC;
+
+const nodeTestHarness: RulesetTestHarness = {
+  setUp: (rules: any, dataLayer: any) => {
+    setupGlobals(
+      Object.keys(dataLayer).map((key) => [key, dataLayer[key]]),
+    );
+    ExpectObserver.getInstance().create({ rules });
+    return Promise.resolve();
+  },
+
+  tearDown: () => {
+    ExpectObserver.getInstance().cleanup();
+    return Promise.resolve();
+  },
+
+  execute: (action: (args: any[]) => void, args?: any[]) => {
+    action(args || []);
+    return Promise.resolve();
+  },
+
+  popEvent: async (timeoutMs: number = 1000) => new Promise<any[]>((resolve, reject) => {
+    const startTime = new Date().getTime();
+
+    const checkForEvents = setInterval(() => {
+      const currentTime = new Date().getTime();
+      if (currentTime - startTime >= timeoutMs) {
+        clearInterval(checkForEvents);
+        reject(new Error(`Timeout ${timeoutMs} exceeded waiting for FS.event call.`));
+      }
+
+      if (expectGlobal('FS').callQueues.event.length) {
+        clearInterval(checkForEvents);
+        resolve(expectFS('event'));
+      }
+    }, 50);
+  }),
+};
+
+class BrowserTestHarness implements RulesetTestHarness {
+  private browser: Browser = null!;
+
+  private page: Page = null!;
+
+  constructor(private readonly browserPromise: Promise<Browser>) {
+    if (!dloScriptSrc) {
+      throw new Error(
+        'PLAYWRIGHT_DLO_SCRIPT_SRC must be configured with a DLO library script location for browser tests.',
+      );
+    }
+  }
+
+  async setUp(rules: any, dataLayer: any) {
+    this.browser = await this.browserPromise;
+    this.page = await this.browser.newPage();
+
+    await this.page.evaluate(([localRules, localDataLayer, localDloScriptSrc]) => {
+      // This allows node and playwright tests to reference the same global/window name.
+      const globalThis: any = window;
```

Comment:
document looks like another global variable, could we define window this way somehow, that way we don't have to stick the `:any` definition on it.

Reply:
I added some global/window typings to broadly reduce `any` usage in tests and within the test env/harness logic. LMKWYT.